File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Ruby/lib/cultist_player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'cultist'
require_relative 'player'

class CultistPlayer < Player
  @@totalCultistPlayers = 0
  def initialize(p,c)
    copy(p)
    @myCultistCard = c
    @@totalCultistPlayers += 1
  end
  
  def combatLevel
    (super*1.7 + @myCultistCard.gainedLevels*@@totalCultistPlayers).floor
  end
  
  def oponentLevel(m)
    m.combatLevelAgainstCultistPlayer
  end
  
  def shouldConvert
    false
  end
  
  def giveMeATreasure
    n = @visibleTreasures.size
    @visibleTreasures[rand(n)]
  end
  
  def canYouGiveMeATreasure
    return !@visibleTreasures.empty?
  end
  
  def self.totalCultistPlayers
    @@totalCultistPlayers
  end
  
  def to_s
    super + ", cultist=true"
  end
  protected :combatLevel, :oponentLevel, :shouldConvert
  private :giveMeATreasure, :canYouGiveMeATreasure
end

File: AÑO 2/programacion_y_diseño_orientado_a_objetos/Practicas/Napakalaki_Java/src/NapakalakiGame/CultistPlayer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.Random;

/**
 *
 * @author juanka1995
 */
public class CultistPlayer extends Player{
    private static int totalCultistPlayers = 0;
    private Cultist myCultistCard;
    
    public CultistPlayer(Player p, Cultist c){
        super(p);
        totalCultistPlayers++;
        this.myCultistCard = c;
    }
    
    @Override
    protected int getCombatLevel(){
        return (int) (super.getCombatLevel() + 0.7*super.getCombatLevel() + myCultistCard.getGainedLevels()*getTotalCultistPlayers());
    }
    
    @Override
    protected int getOponentLevel(Monster m){
        return m.getCombatLevelAgainstCultistPlayer();
    }
    
    @Override
    protected boolean shouldConvert(){
        return false;
    }
    
    @Override
    public boolean isACultistPlayer(){
        return true;
    }
    
    private Treasure giveMeATreasure(){
        Random rand = new Random();
        return this.getVisibleTreasures().get(rand.nextInt(getVisibleTreasures().size())+1);
    }
    
    private boolean canYouGiveMeATreasure(){
        return this.getEnemy().getVisibleTreasures().isEmpty();
    }
    
    public int getTotalCultistPlayers(){
        return totalCultistPlayers;
    }    
    
}
